File: python/src/ancilis/testing/fake_producer.py
"""FakeProducer — inject synthetic evidence without running real agent code."""

from __future__ import annotations

import hashlib
from typing import Any

from ancilis.engine.action import Action
from ancilis.engine.registry import ToolEntry, ToolRegistry
from ancilis.producers.protocol import ProducerType
from ancilis.testing._helpers import make_action


class FakeProducer:
    """Injects synthetic evidence without running real agent code.

    Implements the ActionProducer protocol. Use in tests to produce Action
    objects with controlled data, without any real tool connectivity.

    Usage::

        from ancilis.testing import FakeProducer

        producer = FakeProducer("identity")
        producer.emit("user.id", "alice")
        producer.emit("session.start", {"timestamp": "2026-04-11T10:00:00Z"})

        action = producer.make_action()
        result = engine.evaluate(action)
    """

    def __init__(
        self,
        producer_name: str = "fake",
        agent_id: str = "test-agent",
        agent_owner: str | None = None,
    ) -> None:
        self._producer_name = producer_name
        self._agent_id = agent_id
        self._agent_owner = agent_owner
        self._emitted: dict[str, Any] = {}

    @property
    def producer_type(self) -> ProducerType:
        return ProducerType.MANUAL

    @property
    def producer_version(self) -> str:
        return "0.1.0-test"

    def emit(self, key: str, value: Any) -> None:
        """Record a synthetic evidence item.

        Emitted items are available via ``emitted_data`` and are included
        as parameters in any Action created with ``make_action()``.
        """
        self._emitted[key] = value

    @property
    def emitted_data(self) -> dict[str, Any]:
        """All emitted evidence items (read-only copy)."""
        return dict(self._emitted)

    def clear(self) -> None:
        """Reset all emitted items."""
        self._emitted.clear()

    def make_action(
        self,
        tool_name: str | None = None,
        parameters: dict[str, Any] | None = None,
        session_id: str | None = None,
        data_classifications: list[str] | None = None,
        source_type: str = "agent",
    ) -> Action:
        """Create an Action for engine evaluation.

        Parameters are merged from emitted items (base) plus any explicit
        overrides supplied here.
        """
        merged_params = dict(self._emitted)
        if parameters:
            merged_params.update(parameters)

        return make_action(
            tool_name=tool_name or self._producer_name,
            agent_id=self._agent_id,
            agent_owner=self._agent_owner,
            parameters=merged_params,
            session_id=session_id,
            data_classifications=data_classifications,
            source_type=source_type,
        )

    # --- ActionProducer protocol ---

    def translate(self, raw_invocation: Any) -> Action:
        """Translate a raw dict invocation into an Action."""
        if isinstance(raw_invocation, dict):
            return self.make_action(
                tool_name=raw_invocation.get("tool", self._producer_name),
                parameters=raw_invocation.get("parameters"),
            )
        return self.make_action()

    def compute_tool_hash(self, tool_identifier: Any) -> str:
        return hashlib.sha256(str(tool_identifier).encode()).hexdigest()

    def register_tools(self, registry: ToolRegistry) -> list[str]:
        tool_name = self._producer_name
        registry.register(ToolEntry(name=tool_name))
        return [tool_name]

File: typescript/src/ancilis/testing/fake-producer.ts
/** FakeProducer — inject synthetic evidence for testing without running real agent code. */

import { randomUUID } from "node:crypto";
import { loadConfig } from "../config/index.js";
import type { ResolvedConfig } from "../config/index.js";
import { Engine } from "../engine/engine.js";
import type { Action } from "../engine/action.js";
import type { EvaluationResult } from "../engine/result.js";
import { MockEvidenceStore } from "./mock-evidence-store.js";

export interface FakeEvaluationResult {
  action: Action;
  evaluation: EvaluationResult;
  record: Awaited<ReturnType<MockEvidenceStore["store"]>>;
}

/**
 * FakeProducer runs engine evaluations against a `MockEvidenceStore`,
 * letting tests verify control behavior without real agent code or
 * platform connectivity.
 *
 * @example
 * ```ts
 * const producer = new FakeProducer();
 * const { evaluation } = await producer.evaluate("read_file", { path: "/etc/passwd" });
 * expectControlToPass(evaluation, "PR-01");
 * ```
 *
 * Supply a custom `ResolvedConfig` to test overlay/control combinations:
 * ```ts
 * const producer = new FakeProducer({ config: ComplianceScenarios.missingIdentity() });
 * const { evaluation } = await producer.evaluate("send_email");
 * expectControlToFail(evaluation, "PR-01");
 * ```
 */
export class FakeProducer {
  readonly store: MockEvidenceStore;
  readonly config: ResolvedConfig;
  private _engine: Engine;
  private _defaultAgentId: string | null;

  constructor(options?: {
    config?: ResolvedConfig;
    certifications?: string[];
    /** Force a specific agentId on every evaluate() call (useful for failure scenarios). */
    defaultAgentId?: string | null;
  }) {
    this.config =
      options?.config ??
      loadConfig({ raw: { agent: { name: "test-agent" } } });
    this.store = new MockEvidenceStore({
      certifications: options?.certifications ?? this.config.activeCertifications,
    });
    this._engine = new Engine(this.config);
    this._defaultAgentId = options?.defaultAgentId ?? null;
  }

  /**
   * Evaluate a named tool call and store evidence.
   *
   * @param toolName  Tool / action name (e.g. `"read_file"`, `"send_email"`).
   * @param params    Optional raw parameters for the action.
   * @param agentId   Override the agent identity for this call.
   */
  async evaluate(
    toolName: string,
    params: Record<string, unknown> = {},
    agentId?: string,
  ): Promise<FakeEvaluationResult> {
    const action: Action = {
      actionId: randomUUID(),
      timestamp: new Date().toISOString(),
      agentId: agentId ?? this._defaultAgentId ?? this.config.agentId ?? this.config.agentName,
      sourceType: "fake",
      producerType: "fake",
      producerVersion: "0.0.0",
      agentOwner: this.config.agentOwner ?? null,
      actionType: "tool_call",
      tool: { name: toolName, descriptionHash: null },
      parameters: { raw: params, parameterHash: "fake" },
      context: {
        dataClassifications: [...this.config.dataClassifications.values()].flat(),
        activeOverlays: [...this.config.activeOverlays.keys()],
      },
    };

    const evaluation = this._engine.evaluate(action);
    const record = await this.store.store(evaluation, toolName);
    return { action, evaluation, record };
  }

  /**
   * Run multiple tool calls in sequence and return all results.
   */
  async evaluateAll(
    calls: Array<{ toolName: string; params?: Record<string, unknown>; agentId?: string }>,
  ): Promise<FakeEvaluationResult[]> {
    const results: FakeEvaluationResult[] = [];
    for (const call of calls) {
      results.push(await this.evaluate(call.toolName, call.params, call.agentId));
    }
    return results;
  }

  /** Reset the in-memory store between test cases. */
  reset(): void {
    this.store.clear();
  }
}
